src/misc/cloudAuth.ts

```diff
@@ -0,0 +1,138 @@
+import { Request, Server } from '@hapi/hapi';
+import axios from 'axios';
+import Conf = require('conf');
+import { createHash } from 'crypto';
+import open = require('open');
+import { stringify } from 'querystring';
+import { cpu, mem, osInfo, system } from 'systeminformation';
+import { v4 as uuidv4 } from 'uuid';
+
+const cloudUrl = 'https://cloud-beta.rocket.chat';
+
+export interface ICloudAuthResult {
+    token: string;
+}
+
+export class CloudAuth {
+    private config: Conf;
+    private codeVerifier: string;
+    private server: Server;
+    private redirectUri: string;
+
+    constructor() {
+        this.codeVerifier = uuidv4() + uuidv4();
+    }
+
+    public async executeAuthFlow(): Promise<ICloudAuthResult> {
+        await this.initialize();
+
+        return new Promise((resolve, reject) => {
+            const port = 3005;
+            try {
+                this.redirectUri = `http://localhost:${ port }/callback`;
+
+                this.server = new Server({ host: 'localhost', port });
+                this.server.route({
+                    method: 'GET',
+                    path: '/callback',
+                    handler: async (request: Request) => {
+                        try {
+                            const code = request.query.code;
+                            const token = await this.fetchToken(code);
+
+                            resolve({ token });
+                            return 'Thank you. You can close this tab.';
+                        } catch (err) {
+                            reject(err);
+                        } finally {
+                            this.server.stop();
+                        }
+                    },
+                });
+
+                const codeChallenge = createHash('sha256').update(this.codeVerifier).digest('base64');
+                const authorizeUrl = this.buildAuthorizeUrl(codeChallenge);
+                open(authorizeUrl);
+
+                this.server.start();
+            } catch (e) {
+                // tslint:disable-next-line:no-console
+                console.log('Error inside of the execute:', e);
+            }
+        });
+    }
+
+    public async hasToken(): Promise<boolean> {
+        await this.initialize();
+
+        return this.config.has('our.a.token');
+    }
+
+    public async getToken(): Promise<string> {
+        await this.initialize();
+
+        return this.config.get('our.a.token', '') as string;
```

Comment:
https://github.com/RocketChat/Rocket.Chat/blob/develop/app/cloud/server/functions/getUserCloudAccessToken.js#L39

here is an example of checking if token is still valid and then refreshing it if not.  Temporarily you can use this secret: `5FEKpI-vcjP-sdmhPi2Z` i've still got to make that route handle with out a secret on the pkce clients